### PR TITLE
Added authentication system

### PR DESCRIPTION
**Pull request by rokuz**
https://jira.mail.ru/browse/MAPSME-5500

### CODE REVIEW

File: android/src/com/mapswithme/maps/Framework.java [hunk 1]
```diff
@@ -77,6 +77,13 @@
 
   public static final int ROUTE_REBUILD_AFTER_POINTS_LOADING = 0;
 
+  @Retention(RetentionPolicy.SOURCE)
+  @IntDef({SOCIAL_TOKEN_FACEBOOK, SOCIAL_TOKEN_FACEBOOK})
```
Comment (alexzatsepin) [inline android/src/com/mapswithme/maps/Framework.java]: `@IntDef({SOCIAL_TOKEN_FACEBOOK, SOCIAL_TOKEN_FACEBOOK}) -> @IntDef({ SOCIAL_TOKEN_FACEBOOK, SOCIAL_TOKEN_GOOGLE }) `(второй параметр неверный и пробелы пожалуйста)

File: map/user.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+std::string const kMapsMeTokenKey = "MapsMeToken";
+std::string const kServerUrl = PASSPORT_URL;
+
+std::string AuthenticationUrl(std::string const & socialToken,
+                              User::SocialTokenType socialTokenType)
+{
+  if (kServerUrl.empty())
+    return {};
+
+  std::string socialTokenStr;
```
Comment (alexzatsepin) [inline map/user.cpp]: Ассерт на пустой socialTokenStr и на неизвестный тип токена!?

Comment (rokuz) [reply map/user.cpp]: done

File: map/user.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+private:
+  void SetAccessToken(std::string const & accessToken);
+
+  std::string m_accessToken;
+  mutable std::mutex m_mutex;
+  std::condition_variable m_condition;
+  bool m_needTerminate = false;
+  bool m_authenticationInProgress = false;
+  base::WorkerThread m_workerThread;
```
Comment (ygorshenin) [inline map/user.hpp]: Is it good that each User instance has its own worker thread? Do we have some other threads to share with this class?

File: map/user.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+class User
+{
+public:
+  enum SocialTokenType
+  {
+    Facebook,
+    Google
+  };
+  ~User();
```
Comment (ygorshenin) [inline map/user.hpp]: Blank line between enum and dtor, please.

File: map/user.hpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include "base/worker_thread.hpp"
+
+#include <condition_variable>
+#include <mutex>
+#include <string>
+
+class User
```
Comment (ygorshenin) [inline map/user.hpp]: Comment to this class, note thread-safety, please.

File: map/user.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        }
+        else if (request.ErrorCode() == 403) // Forbidden.
+        {
+          ResetAccessToken();
+          break;
+        }
+      }
+
+      // Wait for some time and retry.
+      std::unique_lock<std::mutex> lock(m_mutex);
```
Comment (ygorshenin) [inline map/user.cpp]: Are you sure it's a good idea to lock mutex for a while in the worker thread and in the same time use it in the public API?

File: map/user.hpp (hunk 3 above)
Comment (rokuz) [reply map/user.hpp]: Обсудили

File: map/user.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      request.SetRawHeader("Accept", "application/json");
+      // TODO: Now passport service uses redirection. If it becomes false, uncomment checking.
+      if (request.RunHttpRequest())// && !request.WasRedirected())
+      {
+        if (request.ErrorCode() == 200) // Ok.
+        {
+          SetAccessToken(ParseAccessToken(request.ServerResponse()));
+          break;
+        }
+        else if (request.ErrorCode() == 403) // Forbidden.
```
Comment (ygorshenin) [inline map/user.cpp]: No need in else after break.

File: map/user.cpp (hunk 6 above)
Comment (ygorshenin) [reply map/user.cpp]: Okay, discussed.

File: map/user.hpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  enum SocialTokenType
+  {
+    Facebook,
+    Google
+  };
+  ~User();
+  void Init();
+
+  void Authenticate(std::string const & socialToken, SocialTokenType socialTokenType);
+  bool IsAuthenticated() const;
```
Comment (igrechuhin) [inline map/user.hpp]: мб возвращать SocialTokenType?

Comment (rokuz) [reply map/user.hpp]: Обсудили, пока не нужно

**Pull request merged**